src/pollen.py

```diff
@@ -0,0 +1,99 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Pollen charm business logic."""
+
+import glob
+import subprocess
+from pathlib import Path
+
+from charms.operator_libs_linux.v0 import apt
+from charms.operator_libs_linux.v1 import systemd
+from charms.operator_libs_linux.v2 import snap
+
+from exceptions import ConfigurationWriteError, InstallError
+
+SNAP_NAME = "gtrkiller-pollen"
+
+
+class PollenService:
+    """Pollen service class."""
+
+    def prepare(self, unit_name, charm_state) -> None:
+        """Install packages and write configuration files.
+
+        Args:
+            unit_name: ¨Pollen charm's unit name.
+            charm_state: Pollen charm's CharmState instance.
+
+        Raises:
+            InstallError: if the packages fail to install
+            ConfigurationWriteError: something went wrong writing the configuration
+        """
+        try:
+            apt.update()
+            snap.add(SNAP_NAME, channel="candidate")
+        except FileNotFoundError as exc:
+            raise InstallError from exc
+        try:
+            subprocess.run(
+                [
+                    "rsync",
+                    f"/var/lib/juju/agents/unit-{unit_name}/charm/files/logrotate.conf",
+                    "/etc/logrotate.d/pollen",
+                ],
+                check=True,
+            )
+            subprocess.run(
+                [
+                    "rsync",
+                    f"/var/lib/juju/agents/unit-{unit_name}/charm/files/rsyslog.conf",
+                    "/etc/rsyslog.d/40-pollen.conf",
+                ],
+                check=True,
+            )
+            systemd.service_restart("rsyslog.service")
+        except (FileNotFoundError, systemd.SystemdError) as exc:
+            raise ConfigurationWriteError from exc
+        if (
+            glob.glob("/dev/tpm*")
+            or Path("/sys/class/misc/hw_random/rng_available").read_text(encoding="utf-8")
+            != "tpm-rng-0"
+        ):
+            try:
+                apt.add_package("rng-tools5")
+                self.check_rng_file(charm_state)
+                systemd.service_restart("rngd.service")
+            except FileNotFoundError as exc:
```

Comment:
Similar to above, under what conditions would `FileNotFound` be raised?

Reply:
In check_rng_file, if the package has not created the rng-tools-debian file, this can throw an error.

Reply:
perhaps worth a comment?